### ISSUE TITLE
PublicApiAnalyzers don't know how to handle members that are conditionally available

### ISSUE THREAD

**Issue by josh-degraw** [dotnet/roslyn-analyzers#2621]
#### Analyzer package
[Microsoft.CodeAnalysis.PublicApianalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.PublicApiAnalyzers/)

#### Package Version

v2.9.3 (Latest)

#### Diagnostic ID

[RS0017](https://github.com/dotnet/roslyn-analyzers/blob/master/src/PublicApiAnalyzers/Microsoft.CodeAnalysis.PublicApiAnalyzers.md) and [RS0016](https://github.com/dotnet/roslyn-analyzers/blob/master/src/PublicApiAnalyzers/Microsoft.CodeAnalysis.PublicApiAnalyzers.md)

#### Repro steps

1. Have code with public member(s) that is/are only applicable in certain condition, e.g.:

In `MyProject.csproj`:

```xml

  <PropertyGroup Condition="$(TargetFramework.StartsWith('net4'))">
    <DefineConstants>SYSTEM_WEB</DefineConstants>
  </PropertyGroup>
```

In `MyClass.cs`:
```cs
public class MyClass
{
#if SYSTEM_WEB
        public int MyInteger { get; set; }
#endif
}
```

2. Set up project to use `PublicApiAnalyzers`

#### Expected behavior
There should be _some_ mechanism to account for features that are only applicable conditionally, e.g. some prefix or section in the `PublicAPI.*.txt` files or something like that.

#### Actual behavior

Either `RS0016` or `RS0017` needs to be disabled/ignored for it to work properly.

**Comment by mavasani** [comment]
This is by design. The correct way to do this is to have one set of public API files per preprocessor directive and ensure the right set is conditionality included in the MSBuild project file.

**Comment by mavasani** [comment]
Note that analyzer does not see conditionally exclude source code- compiler parses it at trivia, like comments, so it is not even possible for any analyzer to analyze conditionally excluded code.

**Comment by josh-degraw** [comment]
@mavasani Okay, thanks for the clarification. Is there a naming convention that needs to be used to allow for having different sets of the files?

**Comment by drewnoakes** [comment]
@mavasani this would be another good addition to the documentation for this analyzer. I'm happy to put it together. Where would it live?

**Comment by mavasani** [comment]
@drewnoakes - https://github.com/dotnet/roslyn-analyzers/blob/master/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md

**Comment by mavasani** [comment]
Re-opening to track we add documentation for this.
